### PR TITLE
feat: use lumberjack for logs + performance optimizations

### PR DESCRIPTION
**Pull request by bietkul**
#### What does this do / why do we need it?
- Use `cache` for RS API response
- Use `cache` to store valid users to avoid bcrypt comparison
- Avoid using `ioutil.readall` that affects the latency
- Use `lumberjack` package to store logs in to file
- The default file size is `500M` after that lumberjack will handle the rotation

I've tested it with `fluent-bit` and also tested the rotation with lumberjack

### CODE REVIEW

File: model/queryid/queryid.go [hunk 1]
```diff
@@ -1,30 +0,0 @@
-package queryid
```
Comment (siddharthlatest) [inline model/queryid/queryid.go:1]: @bietkul Why is this file deleted?

File: plugins/auth/util.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// UserToPasswordCache represents a map of bcrypt validated users
+var UserToPasswordCache = make(map[string]interface{})
+
+// CurrentProcessMutex to stop concurrent writes on map
+var CurrentProcessMutex = sync.RWMutex{}
+
+// SavePassword saved the password in the cache
+func SavePassword(username string, password string) {
+	CurrentProcessMutex.Lock()
+	defer CurrentProcessMutex.Unlock()
```
Comment (siddharthlatest) [inline plugins/auth/util.go]: @bietkul For a single line mutation, avoid using defer. See the reason here: https://gist.github.com/TheHippo/75babf83556dc6f638a4.

File: plugins/logs/logs.go [hunk 3]
```diff
@@ -72,6 +77,18 @@ func (l *Logs) InitFunc() error {
...
 	}
+	filePath := os.Getenv(envLogFilePath)
+	if filePath == "" {
+		log.Warnln(logTag, envLogFilePath+" is not defined log will get stored at ", defaultLogFilePath)
+		filePath = defaultLogFilePath
+	}
+	// configure lumberjack
+	l.lumberjack = lumberjack.Logger{
+		Filename:   filePath,
+		MaxSize:    500,
```
Comment (siddharthlatest) [inline plugins/logs/logs.go]: Let's do 1GB per file.

Comment (siddharthlatest) [inline plugins/logs/logs.go]:
```suggestion
		MaxSize:    1000,
```

File: plugins/logs/middleware.go [hunk 4]
```diff
@@ -187,15 +200,77 @@ func (l *Logs) recordResponse(request *Request, w *httptest.ResponseRecorder, re
...
+		rsRequestBody, err := request.FromContext(ctx)
+		// ignore error to record > 500 status code logs
 		if err != nil {
-			log.Errorln(logTag, "error encountered while parsing the response: ", err)
+			log.Errorln(logTag, "error encountered while reading request body:", err)
 		}
-		// ignore error to record error logs
-		if err == nil {
-			rec.Response.Took = &resBody.Settings.Took
+		marshalled, err := json.Marshal(rsRequestBody)
```
Comment (siddharthlatest) [inline plugins/logs/middleware.go:202]: Conversion to JSON may not be needed. https://stackoverflow.com/a/49007587/1221677

File: plugins/auth/middleware.go [hunk 5]
```diff
@@ -161,11 +161,14 @@ func (a *Auth) basicAuth(h http.HandlerFunc) http.HandlerFunc {
 			{
 				// if the request is made to elasticsearch using user credentials, then the user has to be an admin
 				reqUser := obj.(*user.User)
-				if hasBasicAuth && bcrypt.CompareHashAndPassword([]byte(reqUser.Password), []byte(password)) != nil {
+				// No need to validate if already validated before
+				if hasBasicAuth && !IsPasswordExist(reqUser.Username, password) && bcrypt.CompareHashAndPassword([]byte(reqUser.Password), []byte(password)) != nil {
```
Comment (siddharthlatest) [inline plugins/auth/middleware.go:165]: confirm this part..

**Review by siddharthlatest — CHANGES_REQUESTED** [review]
Some are comments to verify. Some are changes requested.

File: plugins/logs/middleware.go (hunk 4 above)
Comment (bietkul) [reply plugins/logs/middleware.go:202]: @siddharthlatest It doesn't work https://play.golang.org/p/dBjzZLVn9WW

**Review by siddharthlatest — APPROVED** [review]
